src/observers/link_prefetch_observer.js

```diff
@@ -0,0 +1,184 @@
+
+export class LinkPrefetchObserver {
+  started = false
+  delayBeforePrefetching = 100
+  hoverTriggerEvent = "mouseenter"
+  touchTriggerEvent = "touchstart"
+
+  constructor(delegate, eventTarget) {
+    this.delegate = delegate
+    this.eventTarget = eventTarget
+  }
+
+  start() {
+    if (this.started) return
+
+    if (this.eventTarget.readyState === "loading") {
+      this.eventTarget.addEventListener("DOMContentLoaded", this.#enable, { once: true })
+    } else {
+      this.#enable()
+    }
+  }
+
+  stop() {
+    if (!this.started) return
+
+    this.eventTarget.removeEventListener(this.hoverTriggerEvent, this.#tryToPrefetchRequest, {
+      capture: true,
+      passive: true
+    })
+    this.eventTarget.removeEventListener(this.touchTriggerEvent, this.#tryToPrefetchRequest, {
+      capture: true,
+      passive: true
+    })
+    this.eventTarget.removeEventListener("turbo:before-fetch-request", this.#tryToUsePrefetchedRequest, true)
+    this.started = false
+  }
+
+  #enable = () => {
+    this.eventTarget.addEventListener(this.hoverTriggerEvent, this.#tryToPrefetchRequest, {
+      capture: true,
+      passive: true
+    })
+    this.eventTarget.addEventListener(this.touchTriggerEvent, this.#tryToPrefetchRequest, {
+      capture: true,
+      passive: true
+    })
+    this.eventTarget.addEventListener("turbo:before-fetch-request", this.#tryToUsePrefetchedRequest, true)
+    this.started = true
+  }
+
+  #tryToPrefetchRequest = (event) => {
+    if (getMetaContent("turbo-prefetch") !== "true") return
+
+    const target = event.target
+    const isLink = target.matches && target.matches("a[href]:not([target^=_]):not([download])")
+
+    if (isLink && this.#isPrefetchable(target)) {
+      const link = target
+
+      this.prefetchTimeout = setTimeout(() => {
+        this.#startPrefetch(event, link)
+      } , this.delayBeforePrefetching)
+
+      link.addEventListener("mouseleave", this.#cancelPrefetchTimeoutIfAny, {
+        capture: true,
+        passive: true
+      })
+    }
+  }
+
+  #startPrefetch = (event, link) => {
+    const location = getLocationForLink(link)
+    if (this.delegate.canPrefetchAndCacheRequestToLocation(link, location, event)) {
+      this.delegate.prefetchAndCacheRequestToLocation(link, location, this.#cacheTtl)
+    }
+  }
+
+  #cancelPrefetchTimeoutIfAny = () => {
+    if (this.prefetchTimeout) {
+      clearTimeout(this.prefetchTimeout)
+    }
+  }
+
+  #tryToUsePrefetchedRequest = (event) => {
+    if (event.target.tagName !== "FORM" && event.detail.fetchOptions.method === "get") {
+      const cached = prefetchCache.get(event.detail.url.toString())
+
+      if (cached && cached.ttl > new Date()) {
+        // User clicked link, use cache response and clear cache.
+        event.detail.fetchRequest = cached.fetchRequest
+        prefetchCache.clear()
+      }
+    }
+  }
+
+  prepareRequest(request) {
+    const link = request.target
+
+    request.headers["Sec-Purpose"] = "prefetch"
```

Comment:
`Sec-Header` seems to be stripped from the actual request headers sent by `fetch(..)` - could be it's a reserved header in Chromium? See https://github.com/hotwired/turbo/issues/1107

Reply:
Relevant reference is [this](https://www.w3.org/TR/fetch-metadata/#sec-prefix):

> 4.2. The Sec- Prefix
Each of the headers defined in this document is prefixed with Sec-, which makes them all [forbidden response-header names](https://fetch.spec.whatwg.org/#forbidden-response-header-name), and therefore unmodifiable from JavaScript. This will prevent malicious websites from convincing user agents to send forged metadata along with requests, which should give sites a bit more confidence in their ability to respond reasonably to the advertised information.

> These are forbidden so the user agent remains in full control over them.
> [Header names](https://fetch.spec.whatwg.org/#header-name) starting with `Sec-` are reserved to allow new [headers](https://fetch.spec.whatwg.org/#concept-header) to be minted that are safe from APIs using [fetch](https://fetch.spec.whatwg.org/#concept-fetch) that allow control over [headers](https://fetch.spec.whatwg.org/#concept-header) by developers, such as [XMLHttpRequest](https://xhr.spec.whatwg.org/#xmlhttprequest). [[XHR]](https://fetch.spec.whatwg.org/#biblio-xhr

Reply:
* Chromium [is still working](https://chromestatus.com/feature/6247959677108224), on replacing `Purpose: prefetch` with `Sec-Purpose: Prefetch`.

* Firefox [has now](https://bugzilla.mozilla.org/show_bug.cgi?id=1836328) replaced `X-Moz: Prefetch` with `Sec-Purpose: Prefetch`.

So looks like it's just a matter of time before it's sent. As @rik [pointed out](https://github.com/hotwired/turbo/issues/924#issuecomment-1561480326), and @afcapel [too](https://github.com/hotwired/turbo/pull/925#issuecomment-1691788329), using this header instead of a custom one would be more forward looking.

---

By the way, the reference you link mentions _specific_ headers that are forbidden, and `Sec-Purpose` is not in that list.

Here are some specifications that do mention that `Sec-Purpose` with a value of `prefetch` is valid:

* [Fetch Standard](https://fetch.spec.whatwg.org/#sec-purpose-header)
* [WICG spec](https://wicg.github.io/nav-speculation/prefetch.html#sec-purpose-header)

Though you're probably already aware of them, since you made the change to `Sec-Purpose": "prefetch"`.

I think we should continue this conversation in [the issue you opened](https://github.com/hotwired/turbo/issues/1107) though, since it's a very particular topic. And we can follow-up with a PR to change the header if needed.

Reply:
The issues linked for [Firefox](https://bugzilla.mozilla.org/show_bug.cgi?id=1836328) and [Chromium](https://chromestatus.com/feature/6247959677108224) are both related to what header is sent when prefetching _link-rel prefetch requests_ and does not mention any changes in the `fetch` request.

In the [Fetch Standard](https://fetch.spec.whatwg.org/#forbidden-header-name) under the "Forbidden Header Name" section, the following paragraph explicitly mentions that any header starting with `Sec-` should not be allowed to be set by client (eg. via JS), but rather be fully controlled by the user agent:

> A [header](https://fetch.spec.whatwg.org/#concept-header) (name, value) is forbidden request-header if these steps return true:
> ...
> * If name when [byte-lowercased](https://infra.spec.whatwg.org/#byte-lowercase) [starts with](https://infra.spec.whatwg.org/#byte-sequence-starts-with) `proxy-` or `sec-`, then return true.

The linked spec also mentions _why_ these are reserved, even in the future:

> Note: These are forbidden so the user agent remains in full control over them.
> [Header names](https://fetch.spec.whatwg.org/#header-name) starting with `Sec-` are reserved to allow new [headers](https://fetch.spec.whatwg.org/#concept-header) to be minted that are safe from APIs using [fetch](https://fetch.spec.whatwg.org/#concept-fetch) that allow control over [headers](https://fetch.spec.whatwg.org/#concept-header) by developers, such as [XMLHttpRequest](https://xhr.spec.whatwg.org/#xmlhttprequest). [[XHR]](https://fetch.spec.whatwg.org/#biblio-xhr)

As I understand the above this means that:

1. The `Sec-Purpose` header is **intentionally** reserved for user agents only and **intentionally not** allowed to be set by XMLHttpRequest.
2. The Chromium and Firefox changes are unrelated to this, as they change the header being sent on `<link rel=prefetch>`
3. There's no future plan of making `Sec-Purpose` available to XMLHttpRequest, on the other hand any `Sec-` header will not be allowed via XMLHttpRequest.

Reply:
Thanks for sharing additional details, I'd be happy to continue the discussion on the issue you opened. This information would be more valuable there for others that want to chime in on this.